### PR TITLE
kube: cleanup crdwatcher a bit

### PR DESCRIPTION
**Pull request by howardjohn**
~edit: lets do https://github.com/istio/istio/pull/44899 first, then I will rebase this.~

This reworks how we enable optional watchers, which will watch only if the CRD is present.

Today, we do this a few different ways for variuous usages. This standardizes on a single implementation in `kclient`. 

Clients can specify a "Delay" which will result in a client that may exist now or at some future point:

For example:
```go
		wasm := kclient.NewFiltered[*istioclient.WasmPlugin](c, kubetypes.Filter{
			Delayed: crdwatcher.NewDelayedFilter(gvr.WasmPlugin, c.CrdWatcher()),
		})
```

Aside from the filter during creation, usage for this is completely transparent. The kclient will ensure that we only return HasSynced=true once the CRD is ready and we start watching OR the CrdWatcher has finished syncing and the CRD did not exist.

### CODE REVIEW

**Comment by howardjohn** [comment]
@hzxuzhonghu I put a hold sinec I am not 100% sure this is correct yet, going to self-review it a bit more tomorrow. But wanted to put it up for feedback

**Comment by hzxuzhonghu** [comment]
Sorry for the delay i am a little busy with other work. Will take a look later

File: pilot/pkg/config/kube/crdclient/client.go [hunk 1]
```diff
@@ -195,6 +175,22 @@ func (cl *Client) Run(stop <-chan struct{}) {
...
+		return
+	}
+	for _, s := range cl.schemas.All() {
+		// From the spec: "Its name MUST be in the format <.spec.name>.<.spec.group>."
+		name := fmt.Sprintf("%s.%s", s.Plural(), s.Group())
+		if s.IsBuiltin() {
+			handleCRDAdd(cl, name)
+		} else {
+			if !cl.crdWatcher.Known(s.GroupVersionResource()) {
+				cl.logger.Warnf("Skipping CRD %v as it is not present", s.GroupVersionKind())
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/client.go:189]: The log skipping is confusing, maybe CRD xx is not yet present, will be handled latter once it is created

**Review by hzxuzhonghu — COMMENTED** [review]
Generally it looks viable, previously we didnot check crdWatcher.hasSynced at all, which is not right.

File: pkg/kube/kclient/client_test.go [hunk 2]
```diff
@@ -35,6 +42,75 @@ import (
 	"istio.io/istio/pkg/test/util/retry"
 )
 
+func TestSwappingClient(t *testing.T) {
```
Comment (howardjohn) [inline pkg/kube/kclient/client_test.go:45]: @hzxuzhonghu check this out - experimenting with a client that abstracts the CRD watching away.

```
If CRD is already ready, create normal client
If CRD is not ready:
  While CRDWatcher not synced, return HasSynced=false and all empty responses.
  Once CRDWatcher syncs:
     If have CRD: swap to a normal client
     Else, mark HasSynced=true, return all empty responses. Once the CRD is ready, swap to a normal client
```

It isn't 100% there in current iteration but pretty close

Comment (howardjohn) [reply pkg/kube/kclient/client_test.go:45]: tbh not sure this is a good idea but it was fun

Comment (howardjohn) [reply pkg/kube/kclient/client_test.go:45]: No need in our current use cases. We could add it though.

Comment (howardjohn) [reply pkg/kube/kclient/client_test.go:44]: I improved it up a bit more (still needs just cleaning up). Basically now the fact that its a "delayed until CRD is ready" is completely transparent.

File: pkg/kube/kclient/client.go [hunk 3]
```diff
@@ -182,49 +183,101 @@ func New[T controllers.ComparableObject](c kube.Client) Client[T] {
...
+		// The inf() call is responsible for starting the informer
+		inf := getInf()
+		fc := &informerClient[T]{
+			informer:      inf.Informer,
+			startInformer: inf.Start,
+			filter:        filter.ObjectFilter,
+		}
+		inf.Start(stop)
+		// Swap out the dummy client with the full one
+		delayedClient.set(fc)
```
Comment (hzxuzhonghu) [inline pkg/kube/kclient/client.go:270]: We need a log here

File: pkg/kube/kclient/crdwatcher.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// Run starts the controller. This must be called.
+func (c *crdWatcher) Run(stop <-chan struct{}) {
+	select {
+	case <-c.running:
+		// already started, ignore
+	default:
+		c.stop = stop
+		close(c.running)
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/kube/kclient/crdwatcher.go:77]: thanks, added a mutex

**Review by hzxuzhonghu — APPROVED** [review]
LGTM

**Review by hanxiaop — APPROVED** [review]

**Pull request merged**